Core: RSE expression parser, migrate away from session decorators, pass session as a required parameter, add type annotations, minor refactors

rdimaio opened the pull request:
- part of #6989
- fix #8183

rdimaio committed:
fix(Core): RSE expression parser, migrate away from transactional session; #6989

Closes #8183

rdimaio committed:
refactor(Core): add type annotations to core/rse_expression_parser

Issue: #7267

rdimaio committed:
fix(Core): RSE expression parser, check primitive expression match exists before running group to avoid exception

Issue: #8183

rdimaio committed:
style(Core): RSE expression parser, fix variable name typo (parantheses -> parentheses)

Issue: #6989

rdimaio committed:
fix(API): add filter_replicas_by_site method to use site_selector with session provided

Issue: #6989

rdimaio committed:
docs(Core): site_selector, add missing docstring for params

Issue: #6989

rdimaio committed:
fix(API): redirect, use gateway.filter_replicas_by_site instead of core.site_selector

Issue: #6989

rdimaio committed:
refactor(Core): site_selector, rename to filter_replicas_by_site to better reflect what it does

Issue: #6989

rdimaio committed:
fix(Core): rule, assign source_replica_expression to var before parsing it to avoid type analysis error due to .get possibly returning None

Issue: #6989

rdimaio committed:
refactor(Rules): rule, fix type annotation of filter passed to parse_expression, str -> dict

Issue: #7267

rdimaio committed:
fix(Core): atropos, pass session to parse_expression call

Issue: #6989

rdimaio committed:
fix(Core): bb8, pass session to parse_expression calls

Issue: #6989

rdimaio committed:
fix(Core): conveyor, pass session to parse_expression calls

Issue: #6989

rdimaio committed:
fix(Core): dark reaper, pass session to parse_expression calls

Issue: #6989

rdimaio committed:
fix(Core): reaper, pass session to parse_expression calls

Issue: #6989

rdimaio committed:
fix(Core): suspicious replica recoverer, pass session to parse_expression calls

Issue: #6989

rdimaio committed:
fix(Core): transmogrifier, pass session to parse_expression call

Issue: #6989

rdimaio committed:
test(MultiVO): test_multi_vo, pass session to parse_expression calls

Issue: #6989

rdimaio committed:
test(Replicas): test_replica_sorting, pass session to parse_expression calls

Issue: #6989

rdimaio committed:
test(Replicas): test_rse_expression_parser, pass session to parse_expression calls

Issue: #6989

rdimaio committed:
fix(Core): bb8, migrate select_target_rse from read_session, make session required

Issue: #6989

rdimaio committed:
fix(Core): bb8, migrate rebalance_rse from transactional session, make session a required parameter

Issue: #6989

rdimaio committed:
fix(Core): bb8, rebalance_rse, always pass session

Issue: #6989

rdimaio committed:
fix(Core): bb8, parse_expression, always pass session

Issue: #6989

Geogouz commented:
After rebasing, I'm seeing some errors:

https://github.com/Geogouz/rucio/tree/patch-6989-core-rse_expression_parser (Check the `Last commit`).

Basically, the Automatix daemon still calls it without providing the mandatory `session` argument.

Geogouz commented:
Try:
    
```python
  diff --git a/lib/rucio/daemons/automatix/automatix.py b/lib/rucio/daemons/automatix/automatix.py
  index 3afed3ec635e67d4fd44c4bed40c1789cf861a1b..2c06cb7616c7f2bc424838d1fe033b98b09137db 100644
  --- a/lib/rucio/daemons/automatix/automatix.py
  +++ b/lib/rucio/daemons/automatix/automatix.py
  @@ -132,74 +132,77 @@ def automatix(inputfile: str, sleep_time: int, once: bool = False) -> None:
       """
       Creates an automatix Worker that uploads datasets to a list of rses.
   
       :param inputfile: The input file where the parameters of the distribution is set
       :param sleep_time: Thread sleep time after each chunk of work.
       :param once: Run only once.
       """
       run_daemon(
           once=once,
           graceful_stop=graceful_stop,
           executable=DAEMON_NAME,
           partition_wait_time=1,
           sleep_time=sleep_time,
           run_once_fnc=functools.partial(
               run_once,
               inputfile=inputfile,
           ),
       )
   
   
   def run_once(heartbeat_handler: HeartbeatHandler, inputfile: str, **_kwargs) -> bool:
       _, _, logger = heartbeat_handler.live()
   
       try:
           rse_expr = config_get("automatix", "rses")
  -        rses = sorted([rse['rse'] for rse in parse_expression(rse_expr)])
       except (NoOptionError, NoSectionError, RuntimeError):
           logger(logging.ERROR, "Option rses not found in automatix section")
           return True
  -    except InvalidRSEExpression:
  -        logger(logging.ERROR, "Option rses does not contain a valid RSE expression")
  -        return True
       set_metadata = config_get_bool(
           "automatix", "set_metadata", raise_exception=False, default=True
       )
       dataset_lifetime = config_get_int(
           "automatix", "dataset_lifetime", raise_exception=False, default=0
       )
       account = config_get("automatix", "account", raise_exception=False, default="root")
       scope = config_get("automatix", "scope", raise_exception=False, default="test")
   
       client = Client(account=account)
       vo = map_vo(client.vo)  # type: ignore
       filters = {"scope": InternalScope("*", vo=vo)}
       with db_session(DatabaseOperationType.READ) as session:
           scopes = list_scopes(filter_=filters, session=session)
           if InternalScope(scope, vo=vo) not in scopes:
               logger(logging.ERROR, "Scope %s does not exist. Exiting", scope)
               return True
  +        try:
  +            rses = sorted(
  +                [rse['rse'] for rse in parse_expression(rse_expr, session=session)]
  +            )
  +        except InvalidRSEExpression:
  +            logger(logging.ERROR, "Option rses does not contain a valid RSE expression")
  +            return True
   
       logger(logging.INFO, "Getting data distribution")
       probabilities, data = get_data_distribution(inputfile)
       logger(logging.DEBUG, "Probabilities %s", probabilities)
   
       cycle_stopwatch = Stopwatch()
       successes = []
       failures = []
       for rse in rses:
           stopwatch = Stopwatch()
           _, _, logger = heartbeat_handler.live()
           tmpdir = tempfile.mkdtemp()
           logger(logging.INFO, "Running on RSE %s", rse)
           dic = choose_element(probabilities, data)
           metadata = dic["metadata"]
           try:
               nbfiles = dic["nbfiles"]
           except KeyError:
               nbfiles = 2
               logger(
                   logging.WARNING, "No nbfiles defined in the configuration, will use 2"
               )
           try:
               filesize = dic["filesize"]
           except KeyError:
```

rdimaio committed:
fix(Core): automatix, parse_expression, always pass session

Closes: #8183

rdimaio commented:
> Try:
> 
> ```python
>   diff --git a/lib/rucio/daemons/automatix/automatix.py b/lib/rucio/daemons/automatix/automatix.py
>   index 3afed3ec635e67d4fd44c4bed40c1789cf861a1b..2c06cb7616c7f2bc424838d1fe033b98b09137db 100644
>   --- a/lib/rucio/daemons/automatix/automatix.py
>   +++ b/lib/rucio/daemons/automatix/automatix.py
>   @@ -132,74 +132,77 @@ def automatix(inputfile: str, sleep_time: int, once: bool = False) -> None:
>        """
>        Creates an automatix Worker that uploads datasets to a list of rses.
>    
>        :param inputfile: The input file where the parameters of the distribution is set
>        :param sleep_time: Thread sleep time after each chunk of work.
>        :param once: Run only once.
>        """
>        run_daemon(
>            once=once,
>            graceful_stop=graceful_stop,
>            executable=DAEMON_NAME,
>            partition_wait_time=1,
>            sleep_time=sleep_time,
>            run_once_fnc=functools.partial(
>                run_once,
>                inputfile=inputfile,
>            ),
>        )
>    
>    
>    def run_once(heartbeat_handler: HeartbeatHandler, inputfile: str, **_kwargs) -> bool:
>        _, _, logger = heartbeat_handler.live()
>    
>        try:
>            rse_expr = config_get("automatix", "rses")
>   -        rses = sorted([rse['rse'] for rse in parse_expression(rse_expr)])
>        except (NoOptionError, NoSectionError, RuntimeError):
>            logger(logging.ERROR, "Option rses not found in automatix section")
>            return True
>   -    except InvalidRSEExpression:
>   -        logger(logging.ERROR, "Option rses does not contain a valid RSE expression")
>   -        return True
>        set_metadata = config_get_bool(
>            "automatix", "set_metadata", raise_exception=False, default=True
>        )
>        dataset_lifetime = config_get_int(
>            "automatix", "dataset_lifetime", raise_exception=False, default=0
>        )
>        account = config_get("automatix", "account", raise_exception=False, default="root")
>        scope = config_get("automatix", "scope", raise_exception=False, default="test")
>    
>        client = Client(account=account)
>        vo = map_vo(client.vo)  # type: ignore
>        filters = {"scope": InternalScope("*", vo=vo)}
>        with db_session(DatabaseOperationType.READ) as session:
>            scopes = list_scopes(filter_=filters, session=session)
>            if InternalScope(scope, vo=vo) not in scopes:
>                logger(logging.ERROR, "Scope %s does not exist. Exiting", scope)
>                return True
>   +        try:
>   +            rses = sorted(
>   +                [rse['rse'] for rse in parse_expression(rse_expr, session=session)]
>   +            )
>   +        except InvalidRSEExpression:
>   +            logger(logging.ERROR, "Option rses does not contain a valid RSE expression")
>   +            return True
>    
>        logger(logging.INFO, "Getting data distribution")
>        probabilities, data = get_data_distribution(inputfile)
>        logger(logging.DEBUG, "Probabilities %s", probabilities)
>    
>        cycle_stopwatch = Stopwatch()
>        successes = []
>        failures = []
>        for rse in rses:
>            stopwatch = Stopwatch()
>            _, _, logger = heartbeat_handler.live()
>            tmpdir = tempfile.mkdtemp()
>            logger(logging.INFO, "Running on RSE %s", rse)
>            dic = choose_element(probabilities, data)
>            metadata = dic["metadata"]
>            try:
>                nbfiles = dic["nbfiles"]
>            except KeyError:
>                nbfiles = 2
>                logger(
>                    logging.WARNING, "No nbfiles defined in the configuration, will use 2"
>                )
>            try:
>                filesize = dic["filesize"]
>            except KeyError:
> ```

Thanks, addressed in https://github.com/rucio/rucio/pull/8192/changes/8dbc6302540aa7e795d69dbde4ed42d49ae9b4d1. Instead of applying your patch I just changed that line to be within a session context, just to avoid possibly changing the overall functionality of automatix by moving logic around

bari12 reviewed (approved):
I have spent quite some time going through this, and I think it's all sound. @rdimaio thanks for the effort and sorry for the delay!

- **Confidence in review**: High 🟢
- **Confidence in scope**: High 🟢
- **Quality**: Agree 🟢
- **Security**: Agree 🟢
- **Backwards compatibility**: Agree 🟢
- **Testing**: Agree 🟢
- **Documentation**: Agree 🟢

# Notes for merger
I will merge this within the `40` release since it should be fully transparent.
